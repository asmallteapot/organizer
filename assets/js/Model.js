import React from 'react'
import Events from 'ampersand-events'
import AmpersandModel from 'ampersand-model'
import AmpersandCollection from 'ampersand-collection'
import _ from 'lodash'
import moment from 'moment'
import slug from 'slug'
import { DjangoModel, DjangoCollection } from './ModelBase'

export const Activist = DjangoModel.extend({
  urlRoot: '/api/activists/',
  props: {
    address: 'string',
    created: 'string',
    email: 'string',
    name: 'string',
    rank: 'number',
  }
});

export const Signup = DjangoModel.extend({
  urlRoot: '/api/signups/',
  props: {
    state: 'string',
    action: 'string',
  },
  toJSON() {
    return {...this.serialize(), ...{activist: this.activist.url}};
  },
  children: {
    activist: Activist
  }
});

export const SignupCollection = DjangoCollection.extend({
  url: '/api/signups/',
  model: Signup,
})

export const ActivistCollection = DjangoCollection.extend({
  url: '/api/activists/',
  model: Activist
})

export const Action = DjangoModel.extend({
  urlRoot: '/api/actions/',
  props: {
    name: ['string', true, () => "Untitled"],
    date: ['moment', true, () => moment()],
    fields: ['array', false, () => []],
    forms: ['array', false, () => []]
  },
  derived: {
    slug: {
      deps: ['name'],
      type: 'string',
      fn() {
        return slug(this.name);
      }
    }
  },
  collections: {
    signups: SignupCollection,
    //forms: FormCollection,
    //fields: FieldCollection
  }
});

export const ActionCollection = DjangoCollection.extend({
  url: '/api/actions/',
  model: Action
})

export const Form = DjangoModel.extend({
  urlRoot: '/api/forms/',
  props: {
    active: 'boolean',
    description: 'string',
    next_state: 'string',
    title: 'string'
  },
  session: {
    sideloaded: 'boolean'
  },
  collections: {
    //fields: FieldCollection
  },
  sideloadOrFetch() {
    const hasInline = (typeof INLINE_FORM_DATA != 'undefined');
    if (hasInline && INLINE_FORM_DATA.id == this.id) {
      Raven.captureBreadcrumb({
        message: 'Form loaded from sideload cache',
        category: 'action',
        data: INLINE_FORM_DATA,
      });
      this.set(INLINE_FORM_DATA)
      this.sideloaded = true;
    } else {
      this.sideloaded = false;
      return DjangoModel.prototype.fetch.apply(this, arguments)
    }
  },
  children: {
    action: Action
  }
});

export const FormCollection = DjangoCollection.extend({
  url: '/api/forms/',
  model: Form
});

export const SubmissionField = AmpersandModel.extend({
  props: {
    id: 'number',
    value: 'string'
  }
});

export const SubmissionFieldCollection = AmpersandCollection.extend({
  model: SubmissionField
});

export const User = DjangoModel.extend({
  urlRoot: '/api/users/',
  props: {
    id: 'any'
  }
})

export const Submission = DjangoModel.extend({
  props: {
    name: 'string',
    email: 'string',
    address: 'string',
    form: ['object', true, () => new Form()]
  },
  session: {
    errors: ['object', true, () => {}]
  },
  collections: {
    fields: SubmissionFieldCollection
  },
  sync(method, model, options) {
    if (this.form.isNew()) {
      throw new TypeError("form property must have id");
    }
    function parseErrors(response) {
      if (response.status == 400) {
        return response.json()
          .then((body) => {
            this.errors = body.errors || {};
            let err = new Error();
            err.response = response;
            return Promise.reject(err);
          })
      } else if (!response.ok) {
        let err = new Error();
        err.response = response;
        return Promise.reject(err);
      } else {
        this.errors = {};
      }
      return response;
    }
    return DjangoModel.prototype.sync(method, model, options)
      .then(parseErrors.bind(this));
  },
  toJSON() {
    const serialized = this.serialize('props');
    var fieldValues = {}
    _.each(this.fields, field => {
      fieldValues['input_'+field.id] = field.value;
    })
    return {...serialized, ...fieldValues};
  },
  derived: {
    url: {
      deps: ['*'],
      fn() {
        return '/api/forms/'+this.form.id+'/submit_response/'
      }
    }
  }
});

export function bindToState(self, state, propMap) {
  _.each(_.keys(propMap), propName => {
    state.on('change:'+propName, () => {
      var nextState = {};
      nextState[propName] = state[propName];
      self.setState(nextState);
    })
  })
}

export function bindToCollection(self, collection, key) {
  collection.on('add remove change', () => {
    var nextState = {};
    nextState[key] = collection.models;
    self.setState(nextState);
  });
}

export function withState(Component) {
  return class StateBinding extends React.Component {
    constructor(props) {
      super(props);
      Events.createEmitter(this);
      this.onUpdate = _.debounce(this.onUpdate.bind(this), 100);
      this.state = {updateHash: 0};
      this.mounted = false;
    }

    componentDidMount() {
      this.mounted = true;
      this.rebindProps({}, this.props);
    }

    componentWillUnmount() {
      this.mounted = false;
      this.stopListening();
    }

    onUpdate() {
      if (this.mounted) {
        this.setState({updateHash: this.state.updateHash+1});
      }
    }

    rebindProps(oldProps, nextProps) {
      const nextStates    = _.filter(_.values(nextProps), p => p.isState);
      const currentStates = _.filter(_.values(oldProps), p => p.isState);
      const removedStates = _.difference(currentStates, nextStates);
      const addedStates   = _.difference(nextStates, currentStates);

      _.each(removedStates, this.stopListening);
      _.each(addedStates,   s => this.listenTo(s, 'change', this.onUpdate));
    }

    componentWillReceiveProps(nextProps) {
      this.rebindProps(this.props, nextProps);
    }

    render() {
      return <Component {...this.props} />
    }
  }
}
