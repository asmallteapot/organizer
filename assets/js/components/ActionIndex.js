import React from 'react'
import { APIListDataStore } from './RowDataStore'
import StoreBinding from './StoreBinding'
import { Table } from './DataTable'
import { Link } from 'react-router-dom'
import { titles } from '../TitleManager'

export default class ActionIndex extends React.PureComponent {
  constructor(props) {
    super(props);
    this.store = new APIListDataStore('/api/actions/')
    this.store.reload();
  }

  componentDidMount() {
    titles.setSubtitle("Actions");
  }

  render() {
    const columns = [
      {label: "Name",
       value: "name",
       cell: ({row:{id, name}}) => <Link to={`/organize/action/${id}`}>{name}</Link>},
      {label: "Date",
       value: "date"},
      {label: "Signups",
       value: "signups.length"}
    ];
    return (
      <StoreBinding store={this.store}>
        <Table columns={columns} />
      </StoreBinding>
    )
  }
}
