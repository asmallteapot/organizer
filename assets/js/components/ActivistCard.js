import { DragSource } from 'react-dnd'
import React from 'react'
import Gravatar from 'react-gravatar'
import { ContextMenuTrigger, ContextMenu, MenuItem } from 'react-contextmenu'
import _ from 'lodash'
import { Checkbox } from 'react-form'

class ActivistCardBase extends React.Component {
  render() {
    const classes = "card " +
                    (this.props.dragging ? "dragging " : "");
    const draggable = this.props.connectDragSource(
      <div className={classes} key={this.props.activist.cid}>
        <ContextMenuTrigger id={this.props.activist.cid}>
          <div className="right">
            <div className={"rank rank-" + this.props.activist.rank}>{this.props.activist.rank}</div>
            <ContextMenuTrigger showOnClick={true} holdToDisplay={-1} id={this.props.activist.cid}>
              <button className="options"><i className="fa fa-ellipsis-v" /></button>
            </ContextMenuTrigger>
          </div>
          <div className="name">
            <Gravatar size={24} email={this.props.activist.email} />
            {this.props.activist.name}
          </div>
          <div className="email">{this.props.activist.email}</div>
        </ContextMenuTrigger>
      </div>
    )
    return (
      <div className="card-wrap">
        {draggable}
        <ContextMenu id={this.props.activist.cid}>
        </ContextMenu>
      </div>
    )
  }
}

const dragSpec = {
  beginDrag(props, monitor, component) {
    props.onDragging && props.onDragging(true);
    return {activist: props.activist}
  },
}

const collect = (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  dragging: monitor.isDragging()
})

const ActivistDragSource = DragSource("activist", dragSpec, collect);

const ActivistCard = ActivistDragSource(ActivistCardBase);
export default ActivistCard;
