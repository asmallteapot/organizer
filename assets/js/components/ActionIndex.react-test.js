import React from 'react'
import fetchMock from 'fetch-mock'
import ActionIndex, { ActionCard } from './ActionIndex'
import { mount } from 'enzyme'
import ThemeProvider from 'material-ui/styles/MuiThemeProvider'
import { MemoryRouter as Router } from 'react-router-dom'
import { Action, Signup } from '../Model'
import { CardText } from 'material-ui'
import ActivistCard from './ActivistCard'
import injectTapEventPlugin from 'react-tap-event-plugin';

injectTapEventPlugin();

describe('ActionIndex', () => {
  it('renders', () => {
    const component = mount(<Router><ThemeProvider><ActionIndex /></ThemeProvider></Router>);
    expect(component.html()).toMatchSnapshot()
  })
})

describe('ActionCard', () => {
  it('renders blank action', () => {
    const action = new Action();
    const component = mount(<Router><ThemeProvider><ActionCard action={action}/></ThemeProvider></Router>);
    expect(component.html()).toMatchSnapshot()
  })

  it('renders a complete action', () => {
    const action = new Action({
      description: 'Description',
      signups: [
        new Signup()
      ]
    });
    const component = mount(<Router><ThemeProvider><ActionCard action={action}/></ThemeProvider></Router>);
    expect(component.html()).toMatchSnapshot()
  })
})
