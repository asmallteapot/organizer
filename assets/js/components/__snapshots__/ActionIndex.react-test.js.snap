// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionCard renders a complete action 1`] = `
"<div class=\\"card\\" style=\\"color: rgba(0, 0, 0, 0.87); background-color: rgb(255, 255, 255); box-sizing: border-box; font-family: Roboto, sans-serif; box-shadow: 0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12); border-radius: 2px; z-index: 1;\\"><div style=\\"padding-bottom: 0px;\\"><div style=\\"padding: 16px; font-weight: 500; box-sizing: border-box; position: relative; white-space: nowrap; cursor: pointer;\\"><div size=\\"40\\" style=\\"color: rgb(255, 255, 255); background-color: rgb(188, 188, 188); display: inline-flex; font-size: 20px; border-radius: 50%; height: 40px; width: 40px; margin-right: 16px;\\"><!-- react-text: 5 -->1<!-- /react-text --></div><div style=\\"display: inline-block; vertical-align: top; white-space: normal; padding-right: 90px;\\"><span style=\\"color: rgba(0, 0, 0, 0.87); display: block; font-size: 15px;\\">Untitled</span><span style=\\"color: rgba(0, 0, 0, 0.54); display: block; font-size: 14px;\\">a few seconds ago</span></div><button tabindex=\\"0\\" type=\\"button\\" style=\\"border: 10px; box-sizing: border-box; display: inline-block; font-family: Roboto, sans-serif; cursor: pointer; text-decoration: none; margin: auto; padding: 12px; outline: none; font-size: 0; font-weight: inherit; position: absolute; z-index: 1; overflow: visible; width: 48px; height: 48px; top: 0px; bottom: 0px; right: 4px; background: none;\\"><div><svg viewBox=\\"0 0 24 24\\" style=\\"display: inline-block; color: rgba(0, 0, 0, 0.87); fill: currentColor; height: 24px; width: 24px;\\"><path d=\\"M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z\\"></path></svg></div></button></div><div style=\\"padding: 8px; position: relative;\\"><a href=\\"/organize/action/undefined\\" style=\\"margin-right: 8px;\\"><div style=\\"color: rgba(0, 0, 0, 0.87); background-color: rgb(255, 255, 255); box-sizing: border-box; font-family: Roboto, sans-serif; box-shadow: 0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12); border-radius: 2px; display: inline-block; min-width: 88px;\\"><button tabindex=\\"0\\" type=\\"button\\" style=\\"border: 10px; box-sizing: border-box; display: inline-block; font-family: Roboto, sans-serif; cursor: pointer; text-decoration: none; margin: 0px; padding: 0px; outline: none; font-size: inherit; font-weight: inherit; position: relative; z-index: 1; height: 36px; line-height: 36px; width: 100%; border-radius: 2px; background-color: rgb(0, 188, 212); text-align: center;\\"><div><div style=\\"height: 36px; border-radius: 2px; top: 0px;\\"><span style=\\"position: relative; opacity: 1; font-size: 14px; letter-spacing: 0; text-transform: uppercase; font-weight: 500; margin: 0px; padding-left: 16px; padding-right: 16px; color: rgb(255, 255, 255);\\">Open</span></div></div></button></div></a></div><div class=\\"bottom-bar\\"><div class=\\"prospective\\" style=\\"width: 0%;\\"></div><div class=\\"contacted\\" style=\\"width: 0%;\\"></div><div class=\\"confirmed\\" style=\\"width: 0%;\\"></div><div class=\\"attended\\" style=\\"width: 0%;\\"></div><div class=\\"noshow\\" style=\\"width: 0%;\\"></div><div class=\\"cancelled\\" style=\\"width: 0%;\\"></div></div></div></div>"
`;

exports[`ActionCard renders blank action 1`] = `
"<div class=\\"card\\" style=\\"color: rgba(0, 0, 0, 0.87); background-color: rgb(255, 255, 255); box-sizing: border-box; font-family: Roboto, sans-serif; box-shadow: 0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12); border-radius: 2px; z-index: 1;\\"><div style=\\"padding-bottom: 0px;\\"><div style=\\"padding: 16px; font-weight: 500; box-sizing: border-box; position: relative; white-space: nowrap; cursor: pointer;\\"><div size=\\"40\\" style=\\"color: rgb(255, 255, 255); background-color: rgb(188, 188, 188); display: inline-flex; font-size: 20px; border-radius: 50%; height: 40px; width: 40px; margin-right: 16px;\\"><!-- react-text: 5 -->0<!-- /react-text --></div><div style=\\"display: inline-block; vertical-align: top; white-space: normal; padding-right: 90px;\\"><span style=\\"color: rgba(0, 0, 0, 0.87); display: block; font-size: 15px;\\">Untitled</span><span style=\\"color: rgba(0, 0, 0, 0.54); display: block; font-size: 14px;\\">a few seconds ago</span></div><button tabindex=\\"0\\" type=\\"button\\" style=\\"border: 10px; box-sizing: border-box; display: inline-block; font-family: Roboto, sans-serif; cursor: pointer; text-decoration: none; margin: auto; padding: 12px; outline: none; font-size: 0; font-weight: inherit; position: absolute; z-index: 1; overflow: visible; width: 48px; height: 48px; top: 0px; bottom: 0px; right: 4px; background: none;\\"><div><svg viewBox=\\"0 0 24 24\\" style=\\"display: inline-block; color: rgba(0, 0, 0, 0.87); fill: currentColor; height: 24px; width: 24px;\\"><path d=\\"M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z\\"></path></svg></div></button></div><div style=\\"padding: 8px; position: relative;\\"><a href=\\"/organize/action/undefined\\" style=\\"margin-right: 8px;\\"><div style=\\"color: rgba(0, 0, 0, 0.87); background-color: rgb(255, 255, 255); box-sizing: border-box; font-family: Roboto, sans-serif; box-shadow: 0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12); border-radius: 2px; display: inline-block; min-width: 88px;\\"><button tabindex=\\"0\\" type=\\"button\\" style=\\"border: 10px; box-sizing: border-box; display: inline-block; font-family: Roboto, sans-serif; cursor: pointer; text-decoration: none; margin: 0px; padding: 0px; outline: none; font-size: inherit; font-weight: inherit; position: relative; z-index: 1; height: 36px; line-height: 36px; width: 100%; border-radius: 2px; background-color: rgb(0, 188, 212); text-align: center;\\"><div><div style=\\"height: 36px; border-radius: 2px; top: 0px;\\"><span style=\\"position: relative; opacity: 1; font-size: 14px; letter-spacing: 0; text-transform: uppercase; font-weight: 500; margin: 0px; padding-left: 16px; padding-right: 16px; color: rgb(255, 255, 255);\\">Open</span></div></div></button></div></a></div><div class=\\"bottom-bar\\"><div class=\\"prospective\\"></div><div class=\\"contacted\\"></div><div class=\\"confirmed\\"></div><div class=\\"attended\\"></div><div class=\\"noshow\\"></div><div class=\\"cancelled\\"></div></div></div></div>"
`;

exports[`ActionIndex renders 1`] = `
"<div><h1>Actions</h1><div class=\\"action-list\\"><div class=\\"card\\" style=\\"color: rgba(0, 0, 0, 0.87); background-color: rgb(255, 255, 255); box-sizing: border-box; font-family: Roboto, sans-serif; box-shadow: 0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12); border-radius: 2px; z-index: 1;\\"><div style=\\"padding-bottom: 0px;\\"><a href=\\"/organize/action/new\\"><div style=\\"padding: 16px; font-weight: 500; box-sizing: border-box; position: relative; white-space: nowrap;\\"><div size=\\"40\\" style=\\"color: rgb(255, 255, 255); background-color: rgb(188, 188, 188); display: inline-flex; font-size: 20px; border-radius: 50%; height: 40px; width: 40px; margin-right: 16px;\\"><svg viewBox=\\"0 0 24 24\\" style=\\"display: inline-block; color: rgb(255, 255, 255); fill: #ffffff; height: 24px; width: 24px; font-size: 24px; margin: 8px;\\"><path d=\\"M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z\\"></path></svg></div><div style=\\"display: inline-block; vertical-align: top; white-space: normal; padding-right: 90px;\\"><span style=\\"color: rgba(0, 0, 0, 0.87); display: block; font-size: 15px;\\">Create new action</span><span style=\\"color: rgba(0, 0, 0, 0.54); display: block; font-size: 14px;\\"></span></div></div><div style=\\"padding: 16px; font-size: 14px; color: rgba(0, 0, 0, 0.87);\\">Start a new action, get signups.</div></a></div></div></div></div>"
`;
